core/services/workflows/engine.go

```diff
@@ -53,98 +75,114 @@ func (e *Engine) Start(ctx context.Context) error {
-				if s.Ref == keywordTrigger {
-					return nil
-				}
+		// If the capability already exists, that means we've already registered it
+		if s.capability != nil {
+			return nil
+		}
 
-				// If the capability already exists, that means we've already registered it
-				if s.capability != nil {
-					return nil
-				}
+		cp, err := e.registry.Get(ctx, s.Type)
+		if err != nil {
+			return fmt.Errorf("failed to get capability with ref %s: %s", s.Type, err)
+		}
+		// We configure actions, consensus and targets here, and
+		// they all satisfy the `CallbackExecutable` interface
+		cc, ok := cp.(capabilities.CallbackExecutable)
+		if !ok {
+			return fmt.Errorf("could not coerce capability %s to CallbackExecutable", s.Type)
+		}
 
-				cp, innerErr := e.registry.Get(ctx, s.Type)
-				if innerErr != nil {
-					return fmt.Errorf("failed to get capability with ref %s: %s, retrying in %d seconds", s.Type, innerErr, retrySec)
-				}
+		if s.config == nil {
+			configMap, newMapErr := values.NewMap(s.Config)
+			if newMapErr != nil {
+				return fmt.Errorf("failed to convert config to values.Map: %s", newMapErr)
+			}
+			s.config = configMap
+		}
 
-				// We only need to configure actions, consensus and targets here, and
-				// they all satisfy the `CallbackExecutable` interface
-				cc, ok := cp.(capabilities.CallbackExecutable)
-				if !ok {
-					return fmt.Errorf("could not coerce capability %s to CallbackExecutable", s.Type)
-				}
+		registrationRequest := capabilities.RegisterToWorkflowRequest{
+			Metadata: capabilities.RegistrationMetadata{
+				WorkflowID: e.workflow.id,
+			},
+			Config: s.config,
+		}
 
-				if s.config == nil {
-					configMap, ierr := values.NewMap(s.Config)
-					if ierr != nil {
-						return fmt.Errorf("failed to convert config to values.Map: %s", ierr)
-					}
-					s.config = configMap
-				}
+		err = cc.RegisterToWorkflow(ctx, registrationRequest)
+		if err != nil {
+			return fmt.Errorf("failed to register to workflow (%+v): %w", registrationRequest, err)
+		}
 
-				reg := capabilities.RegisterToWorkflowRequest{
-					Metadata: capabilities.RegistrationMetadata{
-						WorkflowID: e.workflow.id,
-					},
-					Config: s.config,
-				}
+		s.capability = cc
+		return nil
+	})
 
-				innerErr = cc.RegisterToWorkflow(ctx, reg)
-				if innerErr != nil {
-					return fmt.Errorf("failed to register to workflow (%+v): %w", reg, innerErr)
-				}
+	return capabilityRegistrationErr
+}
 
-				s.capability = cc
-				return nil
-			})
-			if err != nil {
-				initSuccessful = false
-				e.logger.Error(err)
-			}
+// init does the following:
+//
+//  1. Resolves the underlying capability for each trigger
+//  2. Registers each step's capability to this workflow
+//  3. Registers for trigger events now that all capabilities are resolved
+//
+// Steps 1 and 2 are retried every 5 seconds until successful.
+func (e *Engine) init(ctx context.Context) {
+	defer e.wg.Done()
+	retryMs := 5000
```

Comment:
@HenryNguyen5 xxxRetryMs, xxxMaxRetries don't need to be xxx variables if you add them to the engine Config and substitute the right, production-level defaults in NewEngine.

Thinking ahead, I'm really not a fan of the xxx pattern generally -- I just didn't have a better idea for what to do at the time, but what about adding some lifecycle handlers? This could for example be useful for monitoring, but also for testing to achieve what you want here.

Reply:
Lifecycle handlers would be great, I could start by adding a `postInit` handler that can return either a boolean for indicating success or a lifecycle summary struct. Will likely start with the former to KISS for now